File: php-checks/src/test/resources/checks/phpunit/BooleanOrNullLiteralInAssertionsCheck.php
<?php

use PHPUnit\Framework\TestCase;

class MyTest extends TestCase {
  function testFoo() {
    assertEquals(true, $x); // Noncompliant {{Use assertTrue() instead.}}
    assertEquals(false, $x); // Noncompliant {{Use assertFalse() instead.}}
    assertEquals(null, $x); // Noncompliant {{Use assertNull() instead.}}
    assertEquals('str', $x);

    assertNotEquals(true, $x); // Noncompliant {{Use assertNotTrue() instead.}}
    assertNotEquals(false, $x); // Noncompliant {{Use assertNotFalse() instead.}}
    assertNotEquals(null, $x); // Noncompliant {{Use assertNotNull() instead.}}

    assertSame(true, $x); // Noncompliant {{Use assertTrue() instead.}}
    assertSame(false, $x); // Noncompliant {{Use assertFalse() instead.}}
    assertSame(null, $x); // Noncompliant {{Use assertNull() instead.}}

    assertNotSame(true, $x); // Noncompliant {{Use assertNotTrue() instead.}}
    assertNotSame(false, $x); // Noncompliant {{Use assertNotFalse() instead.}}
    assertNotSame(null, $x); // Noncompliant {{Use assertNotNull() instead.}}

    assertEquals($x, true); // Noncompliant {{Use assertTrue() instead.}}

    assertEquals();
    assertEquals(true);

    assertTrue($x); // Compliant
    assertFalse($x); // Compliant
    assertNull($x); // Compliant
    assertNotNull($x); // Compliant

    assertSame($x, $y); // Compliant
    assertContains(true, [true,false]); // Compliant
  }

  function namedArguments() {
    assertEquals(message: '', actual: $x, expected: 0);
    assertEquals(message: null, actual: $x, expected: 0);
    assertEquals(message: '', actual: $x, expected: null); // Noncompliant {{Use assertNull() instead.}}
  }
}

// For coverage
class Foo {
  public function bar() {
    foo();
  }
}

class Bar extends TestCase {
  public function testBar() {
    $this->assertEquals($x);
  }
}

File: php-checks/src/main/java/org/sonar/php/checks/phpunit/BooleanOrNullLiteralInAssertionsCheck.java
/*
 * SonarQube PHP Plugin
 * Copyright (C) 2010-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
package org.sonar.php.checks.phpunit;

import java.util.Locale;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import org.sonar.check.Rule;
import org.sonar.php.checks.utils.CheckUtils;
import org.sonar.php.checks.utils.PhpUnitCheck;
import org.sonar.php.tree.impl.expression.LiteralTreeImpl;
import org.sonar.php.utils.collections.MapBuilder;
import org.sonar.plugins.php.api.tree.Tree;
import org.sonar.plugins.php.api.tree.declaration.CallArgumentTree;
import org.sonar.plugins.php.api.tree.expression.ExpressionTree;
import org.sonar.plugins.php.api.tree.expression.FunctionCallTree;

@Rule(key = "S2701")
public class BooleanOrNullLiteralInAssertionsCheck extends PhpUnitCheck {
  private static final String MESSAGE = "Use %s instead.";

  private static final Set<String> HANDLED_ASSERTIONS = Set.of(
    "assertEquals",
    "assertSame",
    "assertNotEquals",
    "assertNotSame");

  private static final Set<String> INVERSE_ASSERTIONS = Set.of("assertNotSame", "assertNotEquals");

  private static final Map<String, String> REPLACEMENT_ASSERTIONS = MapBuilder.<String, String>builder()
    .put("true", "assertTrue()")
    .put("false", "assertFalse()")
    .put("null", "assertNull()")
    .put("!true", "assertNotTrue()")
    .put("!false", "assertNotFalse()")
    .put("!null", "assertNotNull()")
    .build();

  @Override
  public void visitFunctionCall(FunctionCallTree tree) {
    if (!isPhpUnitTestCase()) {
      return;
    }

    getAssertion(tree).ifPresent(a -> {
      if (HANDLED_ASSERTIONS.contains(a.name())) {
        verifyAssertion(tree, a.name());
      }
    });
    super.visitFunctionCall(tree);
  }

  private void verifyAssertion(FunctionCallTree tree, String assertionName) {
    findLiteralArgument(tree).ifPresent(l -> suggestAlternative(l, tree, assertionName));
  }

  private void suggestAlternative(LiteralTreeImpl literalTree, FunctionCallTree functionCallTree, String assertionName) {
    String literalValue = literalTree.value().toLowerCase(Locale.ROOT);
    if (INVERSE_ASSERTIONS.contains(assertionName)) {
      literalValue = "!" + literalValue;
    }

    newIssue(functionCallTree, String.format(MESSAGE, REPLACEMENT_ASSERTIONS.get(literalValue)));
  }

  private static Optional<LiteralTreeImpl> findLiteralArgument(FunctionCallTree tree) {
    Optional<CallArgumentTree> expected = CheckUtils.argument(tree, "expected", 0);
    Optional<CallArgumentTree> actual = CheckUtils.argument(tree, "actual", 1);

    if (!expected.isPresent() || !actual.isPresent()) {
      return Optional.empty();
    }

    ExpressionTree expectedValue = expected.get().value();
    ExpressionTree actualValue = actual.get().value();

    if (isLiteral(expectedValue)) {
      return Optional.of((LiteralTreeImpl) expectedValue);
    } else if (isLiteral(actualValue)) {
      return Optional.of((LiteralTreeImpl) actualValue);
    }

    return Optional.empty();
  }

  private static boolean isLiteral(ExpressionTree tree) {
    return tree.is(Tree.Kind.BOOLEAN_LITERAL, Tree.Kind.NULL_LITERAL);
  }
}
